gpytorch/utils/contour_integral_quad.py

```diff
@@ -0,0 +1,137 @@
+    shifts=None,
+    max_lanczos_iter=20,
+    num_contour_quadrature=7,
+    shift_offset=0,
+):
+    r"""
+    Performs :math:`\mathbf K^{1/2} \mathbf b` or `\mathbf K^{-1/2} \mathbf b`
+    using contour integral quadrature.
+
+    .. note::
+        Right now this only works for non-batch matrices
+
+    :param gpytorch.lazy.LazyTensor lazy_tensor: LazyTensor representing :math:`\mathbf K`
+    :param torch.Tensor rhs: Right hand side tensor :math:`\mathbf b`
+    :param bool inverse: (default False) whether to compute :math:`\mathbf K^{1/2} \mathbf b` (if False)
+        or `\mathbf K^{-1/2} \mathbf b` (if True)
+    :param int max_lanczos_iter: (default 10) Number of Lanczos iterations to run (to estimate eigenvalues)
+    :param int num_contour_quadrature: (default 15) How many quadrature samples to use for approximation
+    :rtype: torch.Tensor
+    :return: Approximation to :math:`\mathbf K^{1/2} \mathbf b` or :math:`\mathbf K^{-1/2} \mathbf b`.
+    """
+    import numpy as np
+    from scipy.special import ellipj, ellipk
+
+    output_batch_shape = _mul_broadcast_shape(lazy_tensor.batch_shape, rhs.shape[:-2])
+    preconditioner, preconditioner_lt, _ = lazy_tensor._preconditioner()
+
+    def sqrt_precond_matmul(rhs):
+        if preconditioner_lt is not None:
+            solves, weights, _, _ = contour_integral_quad(preconditioner_lt, rhs, inverse=False)
+            return (solves * weights).sum(0)
+        else:
+            return rhs
+
+    # if not inverse:
+    rhs = sqrt_precond_matmul(rhs)
+
+    if shifts is None:
+        # Determine if init_vecs has extra_dimensions
+        num_extra_dims = max(0, rhs.dim() - lazy_tensor.dim())
+        lanczos_init = rhs.__getitem__(
+            (*([0] * num_extra_dims), Ellipsis, slice(None, None, None), slice(None, 1, None))
+        ).expand(*lazy_tensor.shape[:-1], 1)
+        with warnings.catch_warnings(), torch.no_grad():
+            warnings.simplefilter("ignore", NumericalWarning)  # Supress CG stopping warning
+            _, lanczos_mat = linear_cg(
+                lambda v: lazy_tensor._matmul(v),
+                rhs=lanczos_init,
+                n_tridiag=1,
+                max_iter=max_lanczos_iter,
+                tolerance=1e-5,
+                max_tridiag_iter=max_lanczos_iter,
+                preconditioner=preconditioner,
+            )
+
+        """
+        K^{-1/2} b = 2/pi \int_0^\infty (K - t^2 I)^{-1} dt
+        We'll approximate this integral as a sum using quadrature
+        We'll determine the appropriate values of t, as well as their weights using elliptical integrals
+        """
+
+        # Compute an approximate condition number
+        # We'll do this with Lanczos
+        try:
+            approx_eigs = lanczos_mat.symeig()[0]
+            if approx_eigs.min() <= 0:
+                raise RuntimeError
+        except RuntimeError:
+            approx_eigs = lazy_tensor.diag()
+
+        max_eig = approx_eigs.max(dim=-1)[0]
```

Comment:
should we make these more conserative?

Reply:
I don't know how we would do this in a principled way. For many matrices, our condition number estimate will be accurate to machine precision.

Reply:
why not just do a factor of 2 fudge or something?

Reply:
Just multiplying the max eigenvalue by 2 almost surely gets us further from the true max eigenvalue than the estimate here, and I'm not clear that intentionally overestimating the condition number by a factor of sqrt(2) buys us anything in terms of CIQ convergence? 

These estimates are going to be significantly more accurate than, e.g., 20 iterations of power iteration. They should be pretty darn near exact.